### PR TITLE
fix: update and upgrade pipeline issues

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

1. update pipeline do not change version and tags, closes: #279 .
2. upgrade pipeline will regenerate stack workflow based on the latest parameters and versions, closes: #377.
3. fix retry bugs and support retry when pipeline update/upgrade failed.
4. add new status for pipeline: **Warning**, it's mean pipeline pipeline update/upgrade failed.

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [x] end-to-end tests
  - [x] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/lambda/api/model/pipeline.ts [hunk 1]
```diff
@@ -327,21 +327,15 @@ export class CPipeline {
...
-      await registerClickstreamUser();
-    }
-    // update pipeline tags
-    if (!this.resources || !this.stackTags || this.stackTags?.length === 0) {
-      this.setTags();
-      this.stackTags = this.getStackTags();
-    }
-    const stackTemplateMap = await this.getStackTemplateMap();
+    this.pipeline.workflow = await this.generateWorkflow();
+    this.pipeline.templateVersion = this.resources?.solution?.data.version ?? '';
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/model/pipeline.ts:334]: Will this fix #279?

Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/model/pipeline.ts:335]: fixed.

File: src/control-plane/backend/lambda/api/test/api/pipeline.test.ts [hunk 2]
```diff
@@ -983,35 +987,44 @@ describe('Pipeline test', () => {
...
-              OutputKey: 'Dashboards',
-              OutputValue: '[{"appId":"app1","dashboardId":"clickstream_dashboard_v1_notepad_mtzfsocy_app1"},{"appId":"app2","dashboardId":"clickstream_dashboard_v1_notepad_mtzfsocy_app2"}]',
-            },
-            {
-              OutputKey: 'ObservabilityDashboardName',
-              OutputValue: 'clickstream_dashboard_notepad_mtzfsocy',
-            },
-          ],
+          Outputs: mockOutputs,
+          Tags: [{ Key: 'aws-solution/version', Value: MOCK_SOLUTION_VERSION }],
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/test/api/pipeline.test.ts:1015]: It'd better using constant.

File: src/control-plane/backend/lambda/api/test/api/pipeline.test.ts [hunk 3]
```diff
@@ -1114,7 +1156,8 @@ describe('Pipeline test', () => {
           },
         ],
         templateInfo: {
-          isLatest: false,
+          isLatest: true,
+          pipelineVersion: 'v1.0.0',
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/test/api/pipeline.test.ts:1160]: What's the difference between pipeline version and solution version?

How do we assign then use `isLatest` flag?

**Review by zxkane — CHANGES_REQUESTED** [review]
Do we have new test case or existing test case updating for the fixing #279 ?

File: src/control-plane/backend/lambda/api/service/stack.ts [hunk 4]
```diff
@@ -49,19 +49,25 @@ export class StackManager {
     if (pipeline.workflow) {
       // Deep Copy Workflow
       this.execWorkflow = JSON.parse(JSON.stringify(pipeline.workflow));
+      // this.execWorkflow = { ...pipeline.workflow } as WorkflowTemplate;
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/service/stack.ts]: removed commented

File: src/control-plane/backend/lambda/api/test/api/pipeline.test.ts (hunk 3 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/test/api/pipeline.test.ts:1168]: solution version is the version that save in DDB table.  pipeline version is a attribute of pipeline that set when create/upgrade.

If **isLatest === false** frontend will show Upgrade button.

Comment (zxkane) [reply src/control-plane/backend/lambda/api/test/api/pipeline.test.ts:1168]: So is solution version always the version of web console? is it a volatile one? do you fetch it from the metadata table in runtime?

File: src/control-plane/backend/lambda/api/service/stack.ts [hunk 5]
```diff
@@ -96,28 +102,74 @@ export class StackManager {
...
+    }
+    for (let stackDetail of stackDetails) {
+      if (!stackDetail.stackStatus ||
+        stackDetail.stackStatus?.endsWith('_FAILED') ||
+        stackDetail.stackStatus?.endsWith('_ROLLBACK_COMPLETE') ) {
+        retryStackNames.push(stackDetail.stackName);
+        retryStackTypes.push(stackDetail.stackType);
+      }
+    }
+    if (retryStackTypes.includes(PipelineStackType.INGESTION)) {
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/service/stack.ts:142]: No.

- **updating**: change parameters
- **upgrading**: change template & parameters

**Comment by tyyzqmf** [comment]
> Do we have new test case or existing test case updating for the fixing #279 ?

Added.

File: src/control-plane/backend/lambda/api/test/api/pipeline.test.ts (hunk 3 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/test/api/pipeline.test.ts:1169]: solution version fetch from metadata table in runtime.

**Review by zxkane — APPROVED** [review]

**Pull request merged**